cli/tools/test/fmt.rs

```diff
@@ -74,3 +80,136 @@ pub fn format_test_error(js_error: &JsError) -> String {
     .to_string();
   format_js_error(&js_error)
 }
+
+pub fn format_sanitizer_diff(diff: RuntimeActivityDiff) -> Vec<String> {
+  let mut output = format_sanitizer_accum(diff.appeared, true);
+  output.extend(format_sanitizer_accum(diff.disappeared, false));
+  output.sort();
+  output
+}
+
+fn format_sanitizer_accum(
+  activities: Vec<RuntimeActivity>,
+  appeared: bool,
+) -> Vec<String> {
+  let mut accum = HashMap::new();
+  for activity in activities {
+    let item = format_sanitizer_accum_item(activity);
+    accum.entry(item).or_insert(0).add_assign(1);
+  }
+
+  let mut output = vec![];
+  for ((item_type, item_name), count) in accum.into_iter() {
+    if item_type == RuntimeActivityType::Resource {
+      // TODO(mmastrac): until we implement the new timers and op sanitization, these must be ignored in this path
+      if item_name == "timer" {
+        continue;
+      }
+      let (name, action1, action2) = pretty_resource_name(&item_name);
+      let hint = resource_close_hint(&item_name);
+
+      if appeared {
+        output.push(format!("{name} was {action1} during the test, but not {action2} during the test. {hint}"));
+      } else {
+        output.push(format!("{name} was {action1} before the test started, but was {action2} during the test. \
+          Do not close resources in a test that were not created during that test."));
+      }
+    } else {
+      // TODO(mmastrac): this will be done in a later PR
+      unimplemented!(
+        "Unhandled diff: {appeared} {} {:?} {}",
+        count,
+        item_type,
+        item_name
+      );
+    }
+  }
+  output
+}
+
+fn format_sanitizer_accum_item(
+  activity: RuntimeActivity,
+) -> (RuntimeActivityType, Cow<'static, str>) {
+  let activity_type = activity.activity();
+  match activity {
+    RuntimeActivity::AsyncOp(_, name) => (activity_type, name.into()),
+    RuntimeActivity::Interval(_) => (activity_type, "".into()),
+    RuntimeActivity::Resource(_, name) => (activity_type, name.into()),
+    RuntimeActivity::Timer(_) => (activity_type, "".into()),
+  }
+}
+
+fn pretty_resource_name(
+  name: &str,
+) -> (Cow<'static, str>, &'static str, &'static str) {
+  let (name, action1, action2) = match name {
+    "fsFile" => ("A file", "opened", "closed"),
+    "fetchRequest" => ("A fetch request", "started", "finished"),
```

Comment:
Can you think of a way to make this data be defined on the `Resource` trait - that way we can force via type system to have these descriptions. I'm not sure how we'd look them up though.

Reply:
I think we could add a metadata block for resources but it would require a bit of work to add some registration macros.